dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/implementers/enrollment/AssignValueExecutorTest.java

```diff
@@ -0,0 +1,443 @@
+        attributeA.setUid( ATTRIBUTE_ID );
+        attributeA.setCode( ATTRIBUTE_CODE );
+        attributeA.setValueType( ValueType.NUMBER );
+        dataElementA = createDataElement( 'A' );
+        dataElementA.setUid( DATA_ELEMENT_ID );
+        dataElementA.setCode( DATA_ELEMENT_CODE );
+        ProgramStageDataElement programStageDataElementA = createProgramStageDataElement( firstProgramStage,
+            dataElementA, 0 );
+        firstProgramStage.setProgramStageDataElements( Sets.newHashSet( programStageDataElementA ) );
+        secondProgramStage = createProgramStage( 'B', 0 );
+        secondProgramStage.setValidationStrategy( ValidationStrategy.ON_UPDATE_AND_INSERT );
+        dataElementB = createDataElement( 'B' );
+        dataElementB.setUid( ANOTHER_DATA_ELEMENT_ID );
+        ProgramStageDataElement programStageDataElementB = createProgramStageDataElement( secondProgramStage,
+            dataElementB, 0 );
+        secondProgramStage.setProgramStageDataElements( Sets.newHashSet( programStageDataElementB ) );
+        //        when( preheat.getProgramStage( MetadataIdentifier.ofUid( firstProgramStage ) ) )
+        //            .thenReturn( firstProgramStage );
+        //        when( preheat.getProgramStage( MetadataIdentifier.ofUid( secondProgramStage ) ) )
+        //            .thenReturn( secondProgramStage );
+        //        when( preheat.getDataElement( MetadataIdentifier.ofUid( dataElementA.getUid() ) ) ).thenReturn( dataElementA );
+        when( preheat.getTrackedEntityAttribute( attributeA.getUid() ) )
+            .thenReturn( attributeA );
+        //        when( preheat.getTrackedEntityAttribute( MetadataIdentifier.ofUid( attributeA.getUid() ) ) )
+        //            .thenReturn( attributeA );
+        bundle = TrackerBundle.builder().build();
+        bundle.setPreheat( preheat );
+        when( systemSettingManager.getBooleanSetting( SettingKey.RULE_ENGINE_ASSIGN_OVERWRITE ) )
+            .thenReturn( Boolean.FALSE );
+    }
+
+    @Test
+    void shouldAssignAttributeValueForEnrollmentsWhenAttributeIsEmpty()
+    {
+        when( preheat.getIdSchemes() ).thenReturn( TrackerIdSchemeParams.builder().build() );
+        List<TrackedEntity> trackedEntities = Lists.newArrayList( getTrackedEntitiesWithAttributeNOTSet() );
+        Enrollment enrollmentWithAttributeNOTSet = getEnrollmentWithAttributeNOTSet();
+        List<Enrollment> enrollments = Lists.newArrayList( enrollmentWithAttributeNOTSet );
+        bundle.setTrackedEntities( trackedEntities );
+        bundle.setEnrollments( enrollments );
+
+        AssignValueExecutor executor = new AssignValueExecutor( systemSettingManager,
+            "", TEI_ATTRIBUTE_NEW_VALUE, ATTRIBUTE_ID, enrollmentWithAttributeNOTSet.getAttributes() );
+
+        Optional<ProgramRuleIssue> warning = executor.executeEnrollmentRuleAction( bundle,
+            enrollmentWithAttributeNOTSet );
+
+        Enrollment enrollment = bundle.getEnrollments().stream()
```

Comment:
It looks like you are doing this kind of assertion (the entire block) in multiple tests. Maybe you could make a helper like `assertWarning(...)` to make it easier to figure out what it is that you are asserting without having to follow the stream processing 😄